sbi/inference/snle/snle_base.py

```diff
@@ -295,3 +247,82 @@ def _train(
         # Update summary.
         self._summary["epochs"].append(epoch)
         self._summary["best_validation_log_probs"].append(self._best_val_log_prob)
+
+        # Update TensorBoard and summary dict.
+        self._summarize(
+            round_=self._round, x_o=None, theta_bank=theta, x_bank=x,
+        )
+
+        # Update description for progress bar.
+        if show_train_summary:
+            print(self._describe_round(self._round, self._summary))
+
+        return deepcopy(self._neural_net)
+
+    def build_posterior(
+        self,
+        density_estimator: Optional[TorchModule] = None,
+        mcmc_method: str = "slice_np",
+        mcmc_parameters: Optional[Dict[str, Any]] = None,
+        copy_state_from: Optional[LikelihoodBasedPosteriorType] = None,
+    ) -> LikelihoodBasedPosterior:
+        r"""
+        Build posterior from the neural density estimator.
+
+        SNLE trains a neural network to approximate the likelihood $p(x|\theta)$. The
+        `SNLE_Posterior` class wraps the trained network such that one can directly
+        evaluate the unnormalized posterior log probability
+        $p(\theta|x) \propto p(x|\theta) \cdot p(\theta)$ and draw samples from the
+        posterior with MCMC.
+
+        Args:
+            density_estimator: The density estimator that the posterior is based on.
+                If `None`, use the latest neural density estimator that was trained.
+            mcmc_method: Method used for MCMC sampling, one of `slice_np`, `slice`,
+                `hmc`, `nuts`. Currently defaults to `slice_np` for a custom numpy
+                implementation of slice sampling; select `hmc`, `nuts` or `slice` for
+                Pyro-based sampling.
+            mcmc_parameters: Dictionary overriding the default parameters for MCMC.
+                The following parameters are supported: `thin` to set the thinning
+                factor for the chain, `warmup_steps` to set the initial number of
+                samples to discard, `num_chains` for the number of chains,
+                `init_strategy` for the initialisation strategy for chains; `prior` will
+                draw init locations from prior, whereas `sir` will use
+                Sequential-Importance-Resampling using `init_strategy_num_candidates`
+                to find init locations.
+            copy_state_from: A previous posterior object from which the
+                entire state is copied and then only the `density_estimator` is
+                swapped in. If this is set, all other arguments to this function are
+                overwritten.
+
+        Returns:
+            Posterior $p(\theta|x)$  with `.sample()` and `.log_prob()` methods
+            (the returned log-probability is unnormalized).
+        """
+
+        if density_estimator is None:
+            density_estimator = self._neural_net
```

Comment:
I see, so this `density_estimator` is used only when an external posterior is passed. I would then create a separate method for to avoid confusion. it could even be a method `staticmethod` no? Or which properties of `self` are needed for that?

Reply:
No, the `density_estimator` is always used. It is the neural net that the `posterior` is built from.

I can't exactly follow your suggestion, could you suggest pseudocode?

Reply:
I am just a bit confused about the different scenarios in which one would pass an outside `density_estimator` an outside `posterior` or both. We should discuss this live and with the others I suggest.